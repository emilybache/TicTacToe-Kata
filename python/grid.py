class Grid:
    def __init__(self, width, height):
        self.height = height
        self.width = width
        self.content = {}
        self.comment = ""

    def content_at(self, x, y):
        return self.content.get((x, y), " ")