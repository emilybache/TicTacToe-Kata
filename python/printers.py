

def print_board(board):
    """Prints any grid-like object with content at different x, y co-ordinates.

    The grid-like object must have
    * attribute 'width'
    * attribute 'height'.
    * 'content_at(x, y)' which returns the single character to be printed at that location with co-ordinates (x, y).
    * attribute 'comment' (which may be empty). This is printed below the board.
    """
    field_width = calculate_field_width(board)
    x_coords = [_fixed_width(" ", field_width)] + [_fixed_width(i, field_width) for i in range(board.width)]
    all_content = [x_coords]
    for j in range(board.height):
        y_coords = [_fixed_width(j, field_width)]
        row_contents = y_coords + [_fixed_width(board.content_at(i, j), field_width) for i in range(board.width)]
        all_content.append(row_contents)

    all_rows = [" ".join(row) for row in all_content]
    all_rows.append("")
    board_string = '.\n'.join(all_rows)
    return board_string + board.comment


def calculate_field_width(board):
    return max(len(str(board.width)), len(str(board.height)))


def _fixed_width(s, width):
    "make a string from the object s and pad with whitespace up to width"
    r = str(s)
    r += _whitespace(width - len(r))
    return r


def _whitespace(length):
    "Return a string of whitespace with the requested length"
    return " "*length

