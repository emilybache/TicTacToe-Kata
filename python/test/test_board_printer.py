from approvaltests import verify

from grid import Grid
from printers import print_board


def test_print_empty_board():
    board = Grid(width=3, height=3)
    verify(print_board(board))


def test_print_tic_tac_toe():
    board = Grid(width=3, height=3)
    board.content[(0, 0)] = "x"
    board.content[(2, 0)] = "o"
    board.content[(2, 2)] = "x"
    verify(print_board(board))


def test_print_larger_grid():
    board = Grid(width=12, height=14)
    board.content[(0, 0)] = "x"
    board.content[(2, 0)] = "o"
    board.content[(8, 10)] = "x"
    verify(print_board(board))


def test_print_comment():
    board = Grid(width=3, height=3)
    board.comment = "this is a comment"
    verify(print_board(board))