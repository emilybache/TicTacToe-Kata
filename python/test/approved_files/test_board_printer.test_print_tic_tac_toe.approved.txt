  0 1 2.
0 x   o.
1      .
2     x.
